lib/charms/operator_libs_linux/v0/apt.py

```diff
@@ -1314,6 +1318,203 @@ def _parse(line: str, filename: str) -> DebianRepository:
         else:
             raise InvalidSourceError("An invalid sources line was found in %s!", filename)
 
+    def load_deb822(self, filename: str) -> None:
+        """Load a deb822 format repository source file into the cache.
+
+        Args:
+          filename: the path to the repository file
+
+        In contrast to one-line-style, the deb822 format specifies a repository
+        using a multi-line paragraph. Paragraphs are separated by whitespace,
+        and each definition consists of lines that are either key: value pairs,
+        or continuations of the previous value.
+
+        Read more about the deb822 format here:
+            https://manpages.ubuntu.com/manpages/noble/en/man5/sources.list.5.html
+        For instance, ubuntu 24.04 (noble) lists its sources using deb822 style in:
+            /etc/apt/sources.list.d/ubuntu.sources
+
+        The semantics of `load_deb822` slightly different to `load`:
+            `load` calls `_parse`, which reads a commented out line as an entry that is not enabled
+            `load_deb822` strips out comments entirely when parsing a file into paragraphs, and
+                assumes that comments have been removed when parsing individual paragraphs/entry,
+                instead only reading the 'Enabled' key to determine if an entry is enabled
+        """
+        with open(filename, "r") as f:
+            repos, errors = self._parse_deb822_lines(f, filename=filename)
+
+        for repo in repos:
+            repo_identifier = "{}-{}-{}".format(repo.repotype, repo.uri, repo.release)
+            self._repository_map[repo_identifier] = repo
+
+        if errors:
+            logger.debug(
+                "the following %d error(s) were encountered when reading deb822 format sources:\n%s",
+                len(errors),
+                "\n".join(str(e) for e in errors),
+            )
+
+        if repos:
+            logger.info("parsed %d apt package repositories", len(repos))
+        else:
+            raise InvalidSourceError("all repository lines in '{}' were invalid!".format(filename))
+
+    @classmethod
+    def _parse_deb822_lines(
+        cls,
+        lines: Iterable[str],
+        filename: str = "",
+    ) -> Tuple[List[DebianRepository], List[InvalidSourceError]]:
+        """Parse lines from a deb822 file into a list of repos and a list of errors."""
+        repositories: List[DebianRepository] = []
+        errors: List[InvalidSourceError] = []
+        for paragraph in cls._iter_deb822_paragraphs(lines):
+            try:
+                repos = cls._parse_deb822_paragraph(paragraph, filename=filename)
+            except InvalidSourceError as e:
+                errors.append(e)
+            else:
+                repositories.extend(repos)
+        return repositories, errors
```

Comment:
I know regexes invoke a lot of opinions among software engineers, but I found this regex to be pretty useful for parsing deb822 stanzas. You can see it in action on regex101 - https://regex101.com/r/UriO7l/1:

```python3
_deb822_matcher = re.compile(
    r"""
    (?:Enabled:\s*)?(?P<enabled>.+)?\s?
    Types:\s*(?P<repo_types>.{3,})\s
    URIs:\s*(?P<uris>.+)\s
    Suites:\s*(?P<suites>.+)\s
    (?:Components:\s*)?(?P<components>.+)?\s?
    (?:Signed-By:\s*)?(?P<gpg_key>.+)?\s?
    (?P<options>((.*:\s*)(.+)\s?)*)?
    """,
    re.VERBOSE,
)
```

Then you can glom all the stanzas in a `.sources` file like so:

```python3
for stanza in _deb822_matcher.finditer(content):
    groups = stanza.groupdict()
    enabled = groups.pop("enabled")
    repo_types = groups.pop("repo_types").split()
    uris = groups.pop("uris").split()
    suites = groups.pop("suites").split()
    components = groups.pop("components").split()
    gpg_key = groups.pop("gpg_key")
    raw_options = groups.pop("options")

    if enabled not in ["yes", "no", None]:
        raise InvalidSourceError("...")
    else:
        enabled = True if enabled == "yes" or enabled is None else False

    if len(suites) == 1 and suites[0].endswith("/") and components is not None:
        raise InvalidSourceError("...")
    elif components is None:
        raise InvalidSourceError("...")

    options = {}
    for option in raw_options.splitlines():
        k, v = option.split(":", maxsplit=1)
        options[k] = v.strip()
```

You do lose some of the granularity you're getting such as the line numbers that you're parsing, but there is the added benefit of not needing to maintain a custom file parser. _Trade one form of complexity for another_. The one benefit of the deb822 format is that it's easier for both humans to understand and for machines to manipulate :sweat_smile:

Reply:
Note in case you read the previous comment in emails, I've edited it a bit